Public/Src/Engine/Scheduler/ProcessResourceManager.cs

```diff
@@ -139,6 +140,7 @@ public void TryFreeResources(int requiredRamMb)
             }
 
             using (scope)
+            using (new StoppableTimer(() => scope.RefreshRamUsage(), 0, 2000))

```

Comment:
>using (new StoppableTimer(() => scope.RefreshRamUsage(), 0, 2000)) [](start = 12, length = 66)

Why is this needed?

Reply:
There is no way to get the memory usage of the child processes after they die. That's why, we iterate through the child processes every 2 seconds to get the total workingset usage.

Reply:
We need to iterate through the child processes every 2 seconds to get the peak memory usages. When the child process dies, there is no way to get the memory usage. That's why, we do sampling here. If there is a child process created and terminated under 2 seconds, we might not catch the memory usage of that process.

---
In reply to: [337223679](https://github.com/microsoft/BuildXL/pull/1076#discussion_r337223679) [](ancestors = 337223679)

Reply:
Doesn't the schedule already query the ram usage every 2 seconds?

---
In reply to: [338328220](https://github.com/microsoft/BuildXL/pull/1076#discussion_r338328220) [](ancestors = 338328220,337223679)

Reply:
Not a big deal, but how about having the scheduler do this as a part of UpdateResourceAvailability. I'm slightly concerned about having so many timers.

---
In reply to: [338328220](https://github.com/microsoft/BuildXL/pull/1076#discussion_r338328220) [](ancestors = 338328220,337223679)